Update on UDP policing section (#305)
quicwg/ops-drafts: draft-ietf-quic-manageability.md

accepts QUIC. Sudden drops afterwards will result in slow and costly timeouts
before abandoning the connection.

## UDP Blocking or Throttling {#sec-udp-1312}

Today, UDP is the most prevalent DDoS vector, since it is easy for compromised
non-admin applications to send a flood of large UDP packets (while with TCP the
attacker gets throttled by the congestion controller) or to craft reflection and
amplification attacks. Some networks therefore block UDP traffic.
With increased deployment of QUIC, there is also an increased need to allow
UDP traffic on ports used for QUIC. However, if UDP is generally enabled on these
ports, UDP flood attacks may also used the same ports. One possible
response to this threat is to throttle UDP traffic on the network, allocating a
fixed portion of the network capacity to UDP and blocking UDP datagrams over that
cap. As the portion of QUIC traffic compared to TCP is also expected to
increase over time, limits might need to be adopted dynamically.

> Further, if UDP traffic is throttle, it is reecommended to rather block

Author: ```suggestion
Further, if UDP traffic is throttle, it is recommended to rather block
```

QUIC flows entirely rather than dropping packets randomly. When the handshake is
blocked, QUIC-capable applications may failover to TCP

> (at least applications using well-known UDP ports). However, blocking a

Reviewer: ```suggestion
However, blocking a
```

Well-known ports has nothing to do with it.

random fraction of QUIC packets across 4-tuples will allow many QUIC handshakes
to complete, preventing a TCP failover, but the connections will suffer from

> severe packet loss (see also {{sec-filtering}}). Per-flow throttling can be

Author: hm, it's blocking some flow to throttle the aggregate UDP traffic...

realized based on the hash of the UDP datagram's source and destination

> addresses and blocking a portion of the hash space that corresponds to the

Reviewer: This bit about hash space is unnecessarily narrow.  The overall point is sound: block a subset of flows, not a subset of packets.  Talking about hashing is a distraction.

Author: I disagree. This is about guidance to network function. The more specific we can be, the more likely people do the right thing.

Reviewer 2: I think you're both right. We need an example of how to do efficient per-flow policing here, but it could be Even More Clear that it's an example.

> fraction of UDP traffic one wishes to drop.

Reviewer 3: If we're going to advise this, I think we should be clear that addresses may change for QUIC connections, and when they do, this will result in post-handshake blackholing.

Author: It could but that is a general problem on the Internet today. If the address changes it is not guaranteed that the traffic is still routed correctly...

Reviewer 3: True, but now we have a potentially common case where a NAT rebinds and there's a chance the flow is suddenly blackholed.

Reviewer 3: I think so.  I still think it's worth a wider discussion, but the trade-offs are clearer now.

> 

Author: ```suggestion
While QUIC endpoints are often able to survive address changes, e.g. by NAT rebindings,
blocking a portion of the traffic based on 5-tuple hashing increases the risk of
blackholing an active connection when the address changes.

```

Reviewer: > QUIC connections can be resilient to address changes, such as those caused by NAT rebinding.  However, blocking a portion of flows as identified by address tuples means that some active flows will fail when addresses change.

"blackholing" or black hole is not a useful term to use here.  I would rather you avoid it.

Author: Do you have a better term?

accepts QUIC. Sudden drops afterwards will result in slow and costly timeouts
before abandoning the connection.

## UDP Blocking or Throttling {#sec-udp-1312}

Today, UDP is the most prevalent DDoS vector, since it is easy for compromised
non-admin applications to send a flood of large UDP packets (while with TCP the
attacker gets throttled by the congestion controller) or to craft reflection and
amplification attacks. Some networks therefore block UDP traffic.
With increased deployment of QUIC, there is also an increased need to allow
UDP traffic on ports used for QUIC. However, if UDP is generally enabled on
these ports, UDP flood attacks may also use the same ports. One possible
response to this threat is to throttle UDP traffic on the network, allocating a
fixed portion of the network capacity to UDP and blocking UDP datagrams over
that cap. As the portion of QUIC traffic compared to TCP is also expected to

> increase over time, limits might need to be adopted dynamically.

Author: ```suggestion
increase over time, using such a limit is not recommended but if done,
limits might need to be adopted dynamically.
```

Reviewer 4: Glad we are of the same mind (that a UDP cap is not ideal, but sometimes it may be necessary and if so here's how to do it).
My only comment on this sentence is to ask whether "adopted" should be "adapted". I would say that adoption means the process of starting to apply something, whereas adapted means changing it as you go. It's still possible to say you can adopt something in a dynamic fashion, but "adapted dynamically" is probably clearer.

> Further, if UDP traffic is throttled, it is recommended to block individual

Author: ```suggestion
Further, if UDP traffic is desired to be throttled, it is recommended to block individual
```

Reviewer 4: Looks good.

QUIC flows entirely rather than dropping packets randomly. When the handshake is
blocked, QUIC-capable applications may failover to TCP
However, blocking a
random fraction of QUIC packets across 4-tuples will allow many QUIC handshakes
to complete, preventing a TCP failover, but the connections will suffer from
severe packet loss (see also {{sec-filtering}}). Therefore UDP throttling
should be rather realised by per-flow policing, e.g, based on the hash
of the UDP datagram's source and destination
addresses and blocking a portion of the hash space that corresponds to the
fraction of UDP traffic one wishes to drop.

> While QUIC endpoints are often able to survive address changes, e.g. by NAT rebindings,
> blocking a portion of the traffic based on 5-tuple hashing increases the risk of
> blackholing an active connection when the address changes.

Author: ```suggestion
While QUIC endpoints are often able to survive address changes, e.g. by NAT
rebindings, blocking a portion of the traffic based on 5-tuple hashing increases
the risk of black-holing an active connection when the address changes.
```

accepts QUIC. Sudden drops afterwards will result in slow and costly timeouts
before abandoning the connection.

## UDP Blocking or Throttling {#sec-udp-1312}

Today, UDP is the most prevalent DDoS vector, since it is easy for compromised
non-admin applications to send a flood of large UDP packets (while with TCP the
attacker gets throttled by the congestion controller) or to craft reflection and
amplification attacks. Some networks therefore block UDP traffic.
With increased deployment of QUIC, there is also an increased need to allow
UDP traffic on ports used for QUIC. However, if UDP is generally enabled on
these ports, UDP flood attacks may also use the same ports. One possible
response to this threat is to throttle UDP traffic on the network, allocating a
fixed portion of the network capacity to UDP and blocking UDP datagrams over
that cap. As the portion of QUIC traffic compared to TCP is also expected to
increase over time, using such a limit is not recommended but if done,
limits might need to be adapted dynamically.

> Further, if UDP traffic is desired to be throttled, it is recommended to block individual

Reviewer 2: ```suggestion
Further, if UDP traffic is desired to be throttled, it is recommended to
block individual
```

QUIC flows entirely rather than dropping packets randomly. When the handshake is
blocked, QUIC-capable applications may failover to TCP
However, blocking a
random fraction of QUIC packets across 4-tuples will allow many QUIC handshakes
to complete, preventing a TCP failover, but the connections will suffer from

> severe packet loss (see also {{sec-filtering}}). Therefore UDP throttling
> should be rather realised by per-flow policing, e.g, based on the hash
> of the UDP datagram's source and destination
> addresses and blocking a portion of the hash space that corresponds to the
> fraction of UDP traffic one wishes to drop.

Reviewer 2: ```suggestion
severe packet loss (see also {{sec-filtering}}). Therefore UDP throttling
should be realized by per-flow policing as opposed to per-packet
policing. Note that this per-flow policing should be stateless to avoid
problems with stateful treatment of QUIC flows (see {{sec-stateful}}),
for example blocking a portion of the space of values of a hash function
over the addresses and ports in the UDP datagram.
```
